### ISSUE TITLE
Search just returns "Could not load Inbox :'-("

### ISSUE THREAD

**Issue by Bugsbane** [nextcloud/mail#334]
### Steps to reproduce
1. With mail app open search for any phrase (including ones visibly in mail subjects)

2.Wait for "Searching for {searchphrase}" throbber to disappear

### Expected behaviour
A list of emails containing the search phrase

### Actual behaviour
I just get the error message "Could not load Inbox :'-("

### Mail app

**Mail app version:** 
0.6.2

**Mailserver or service:**
HostGator

**Transport security - IMAP:**
SSL/TLS

**Transport security - SMTP:**
SSL/TLS

**Number of accounts:**
1

**Mail app version build date:** (only if you are using a Nightly Build)
N/A

### Server configuration
**Operating system**:
Debian 8

**Web server:**
Nginx

**Database:**
MariaDB

**PHP version:**
5.6.30

**Version:** (see admin page)
Nextcloud 11.0.2

**Updated from an older version or fresh install:**
Updated since NC 10

**List of activated apps:**

Activity
Auditing/Logging
Collaborative tags
Comments
Deleted files
File sharing
First run wizard
Gallery
LDAP user and group backend
Log Reader
Mail template editor
Nextcloud announcements
Notifications
Password policy
PDF viewer
Server info
Share by mail
Text editor
Theming editor
Update notification
Usage survey
Versions
Video player
Bookmarks
Calendar
Contacts
Mail
News
Notes
Two Factor TOTP Provider
Video calls

**The content of config/config.php:**

<details>
<?php
$CONFIG = array (
  'passwordsalt' => 'XXXX',
  'secret' => 'XXXX',
  'trusted_domains' => 
  array (
    0 => 'localhost',
    1 => 'XXXX.com',
  ),
  'datadirectory' => '/home/XXXX/data',
  'overwrite.cli.url' => 'https://XXXX.com',
  'dbtype' => 'mysql',
  'version' => '11.0.2.7',
  'dbname' => 'XXXX',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'XXXX',
  'dbuser' => 'XXXX',
  'dbpassword' => 'XXXX',
  'logtimezone' => 'UTC',
  'installed' => true,
  'instanceid' => 'oc3seonmlmh9',
  'ldapIgnoreNamingRules' => false,
  'memcache.local' => '\\OC\\Memcache\\APCu',
  'memcache.locking' => '\\OC\\Memcache\\APC',
  'theme' => '',
  'loglevel' => 2,
  'maintenance' => false,                                                                                                                      
  'appstore.experimental.enabled' => true,                                                                                                     
  'mail_from_address' => 'XXXX',                                                                                                              
  'mail_smtpmode' => 'php',                                                                                                                    
  'mail_domain' => 'XXXX.com',                                                                                                                
  'updatechecker' => true,                                                                                                                     
  'logout_url' => 'https://XXXX.com/yunohost/sso/?action=logout',                                                                   
  'htaccess.RewriteBase' => '/',
  'updater.server.url' => 'https://updates.nextcloud.org/updater_server/',
  'updater.release.channel' => 'stable',
  'preview_libreoffice_path' => '/usr/lib/libreoffice/program/soffice',
  'ldapProviderFactory' => '\\OCA\\User_LDAP\\LDAPProviderFactory',
);
</details>

#### Client configuration
**Browser:**
Chromium

**Operating system:**
Ubuntu 16.10

##### Web server error log
```
Nothing in Nginx's error.log
```

##### Server log (data/nextcloud.log)
<details>
```
Nothing relevant (just some HTML purifier logs from News app, which have been found to be an error in HTMLpurifier)
```
</details>

##### Horde IMAP log (data/horde_imap.log)
<details>
File doesn't seem to exist.
</details>

##### Horde SMTP log (data/horde_smtp.log)
<details>
File doesn't seem to exist.
</details>

##### Browser log
<details>
```
Insert your browser log here, this could for example include:

a)
Failed to load resource: the server responded with a status of 504 (Gateway Time-out)
mail.min.js?v=0f0f7ee…:148 [ui] "content:error" 

b) Notihng but a font being blocked by CSP

</details>

**Comment by ChristophWurst** [comment]
> File doesn't seem to exist.

You have to enable IMAP logging, see https://github.com/nextcloud/mail#debug-mode.

Could you please check the browser's network tab? Does the request fail with HTTP 500 or something different?

**Comment by Bugsbane** [comment]
Ok. I've enabled IMAP logging (I'm assuming that the 'app.mail.imaplog.enabled' => true goes in Nextcloud's config.php) and searched again.

Currently, everything in the network tab comes up as 200's except for the messages?filter= queries that sit in pending until they come back as a 504 gateway timeout.

I tried also adding 'app.mail.imap.timeout' => 60, to no effect.

I know that elsewhere when I was doing a very large amount of file syncs with the ownCloud desktop sync client, that it was sometimes hitting problems with ram reaching the php.ini limit and thus some Nextcloud process crashing. My email encompasses some gigabytes, so I'm wondering if something similar is happening here.

**Comment by ChristophWurst** [comment]
Did you search inside the unified inbox? Or maybe a very large mailbox? That could be a reason for the timeout.

**Comment by Bugsbane** [comment]
Well, I only have one account hooked up, but yes, my inbox is pretty large. My server has about 6gb of RAM dedicated to just Nextcloud though, so I would hope it should be possible to configure to work.

At the end of the day though, I would expect at the very least a more graceful fail, like just filtering the emails already shown in the current view (as K9 on Android does) or processing results in pages.

**Comment by ChristophWurst** [comment]
From what I've learning about IMAP while working on this app, searching can be PITA because the SORT capability is required to efficiently filter and sort mailboxes via the IMAP protocol.

What we currently do if SORT is unavailable is to fetch a list of messages, including their UID and date. Then we build the page (by sorting the result by date and taking the top n, possible with a offset). When we have the page (basically a list of UIDs), we load the full message headers with a second IMAP command.

Of course, if anybody knows a more efficient algorithm, I'm happy to discuss improvments. Right now, I don't see any solution to this problem. If you have a large mailbox, this algorithm obviously has its limitations – just image a mailbox with 200k messages and you search for "a" -> even if you "just" fetch UIDs, the response will be *huge*.

I'm closing this ticket now as 'wontfix'. The only realistic improvement we look into would be the local filtering as you already suggested. #326 will make that a bit easier since we're using a bit more intelligent data structures on the client-side. Feel free to open a enhancement ticket for this.

Thanks

**Comment by Bugsbane** [comment]
I wonder how projects like Roundcube / Rainloop handle this? I don't recall ever having any issues with search when using Roundcube on the same email account before.
